osu.Game/Screens/Play/HUD/FumoStrainGraph.cs

```diff
@@ -0,0 +1,216 @@
+
+        private bool useBackgroundGradient = true;
+        private Colour4 backgroundColour = FumoColours.SeaBlue.Light;
+        private Colour4 lineColour = Color4.White;
+
+        private const int highest_point = 5;
+
+        private readonly Box background;
+        private readonly SliderPath path = new SliderPath();
+        private readonly SmoothPath drawablePath;
+
+        private IEnumerable<HitObject>? objects;
+        private float[] values = new float[10];
+
+        public IEnumerable<HitObject> Objects
+        {
+            set
+            {
+                objects = value;
+                refresh();
+            }
+        }
+
+        public FumoStrainGraph()
+        {
+            AutoSizeAxes = Axes.Both;
+
+            InternalChildren = new Drawable[]
+            {
+                background = new Box
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = useBackgroundGradient
+                        ? ColourInfo.GradientVertical(Color4.Black.Opacity(0), backgroundColour.Opacity(0.9f))
+                        : backgroundColour.Opacity(0.9f),
+                    Alpha = 0,
+                    Depth = float.MaxValue,
+                },
+                drawablePath = new SmoothPath
+                {
+                    Anchor = Anchor.BottomLeft,
+                    Origin = Anchor.BottomLeft,
+                    PathRadius = 2,
+                    Colour = lineColour,
+                },
+            };
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            background.FadeTo(0.3f, 200, Easing.In);
+        }
+
+        private void refresh()
+        {
+            if (objects == null || !objects.Any())
+                return;
+
+            values = new float[displayGranularity];
+
+            (double firstHit, double lastHit) = BeatmapExtensions.CalculatePlayableBounds(objects);
+
+            if (lastHit == 0)
+                lastHit = objects.Last().StartTime;
+
+            double interval = (lastHit - firstHit + 1) / displayGranularity;
+
+            foreach (var h in objects)
+            {
+                double endTime = h.GetEndTime();
+
+                Debug.Assert(endTime >= h.StartTime);
+
+                int startRange = (int)((h.StartTime - firstHit) / interval);
+                int endRange = (int)((endTime - firstHit) / interval);
+                for (int i = startRange; i <= endRange; i++)
+                    values[i]++;
+            }
+
+            standardizeValues();
+            updateGraph();
+        }
+
+        private void standardizeValues()
+        {
+            float minValue = values.Min();
+            float maxValue = values.Max() - minValue;
+
+            for (int i = 0; i < displayGranularity; i++)
+            {
+                values[i] = (values[i] - minValue) / maxValue * highest_point;
+            }
```

Comment:
```suggestion
            for (int i = 0; i < displayGranularity; i++)
            {
                values[i] = (values[i] - minValue) / maxValue * highest_point;
            }

            int smoothingPasses = 2;

            for (int pass = 0; pass < smoothingPasses; pass++)
            {
                float[] smoothedValues = new float[displayGranularity];

                for (int i = 0; i < displayGranularity; i++)
                {
                    if (i == 0 || i == displayGranularity - 1)
                    {
                        smoothedValues[i] = values[i];
                        continue;
                    }

                    smoothedValues[i] = 0.25f * values[i-1] + 0.5f * values[i] + 0.25f * values[i+1];
                }

                values = smoothedValues;
            }
```

Having too many scattered sampling points can sometimes be viewed as noise, leading to readability issues. To address this I implemented a smoothing algorithm on the sampling points twice. What do you think? Does it look better while still maintaining the overall trend for the beatmap strain distribution?

Reply:
The smoothing method turns out to be pretty good. I'll push this with some comments soon.

| Standardization method | Graph |
| :-: | :-: |
| Normalization | ![Normalization](https://github.com/user-attachments/assets/ba2725d7-992c-4690-a2b1-d80a18d0e4b2) |
| Normalization (2 passes) | ![Normalizationx2](https://github.com/user-attachments/assets/bddb42da-b2c6-458b-9edd-7495aec2abc6) |
| Normalization + Smoothing | ![Normalization+Smoothing](https://github.com/user-attachments/assets/89332573-083f-440a-ae0c-0668838e5f48) |

Reply:
Personally I prefer Normalization + Smoothing

Reply:
> Personally I prefer Normalization + Smoothing

Yeah that's applied to the latest code. After fixing the tests I'll push it out.